### PR TITLE
chore(types): better types for ProtractorPlugin

### PR DESCRIPTION
**Pull request by thorn0**

### CODE REVIEW

File: lib/plugins.ts [hunk 1]
```diff
@@ -129,7 +129,7 @@ export interface ProtractorPlugin {
    *     for the promise to resolve before continuing.  If the promise is
    *     rejected, a failed assertion is added to the test results.
    */
-  onPageStable?: () => q.Promise<any>;
+  onPageStable?(): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These actually return webdriver promises.  Could you update the type/comment?

File: lib/plugins.ts [hunk 2]
```diff
@@ -113,7 +113,7 @@ export interface ProtractorPlugin {
    *     for the promise to resolve before continuing.  If the promise is
    *     rejected, a failed assertion is added to the test results.
    */
-  onPageLoad?: () => q.Promise<any>;
+  onPageLoad?(): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These actually return webdriver promises.  Could you update the type/comment?

File: lib/plugins.ts [hunk 3]
```diff
@@ -98,7 +98,7 @@ export interface ProtractorPlugin {
    *     will *not* wait before executing the next test, however.  If the promise
    *     is rejected, a failed assertion is added to the test results.
    */
-  postTest?: (passed: boolean, testInfo: any) => q.Promise<any>;
+  postTest?(passed: boolean, testInfo: any): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These can actually return any kind of promise.  How about `void|q.IPromise`?  Also, update the comment to have a return type `{Promise=}`.

File: lib/plugins.ts [hunk 4]
```diff
@@ -79,7 +79,7 @@ export interface ProtractorPlugin {
    *     for the promise to resolve before continuing.  If the promise is
    *     rejected, an error is logged to the console.
    */
-  postResults?: () => q.Promise<any>;
+  postResults?(): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These can actually return any kind of promise.  How about `void|q.IPromise`?  Also, update the comment to have a return type `{Promise=}`.

File: lib/plugins.ts [hunk 5]
```diff
@@ -65,7 +65,7 @@ export interface ProtractorPlugin {
    *     for the promise to resolve before continuing.  If the promise is
    *     rejected, a failed assertion is added to the test results.
    */
-  teardown?: () => q.Promise<any>;
+  teardown?(): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These can actually return any kind of promise.  How about `void|q.IPromise`?  Also, update the comment to have a return type `{Promise=}`.

File: lib/plugins.ts [hunk 6]
```diff
@@ -50,7 +50,7 @@ export interface ProtractorPlugin {
    *     for the promise to resolve before continuing.  If the promise is
    *     rejected, a failed assertion is added to the test results.
    */
-  onPrepare?: () => q.Promise<any>;
+  onPrepare?(): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These can actually return any kind of promise.  How about `void|q.IPromise`?  Also, update the comment to have a return type `{Promise=}`.

File: lib/plugins.ts [hunk 7]
```diff
@@ -34,7 +34,7 @@ export interface ProtractorPlugin {
    *     for the promise to resolve before continuing.  If the promise is
    *     rejected, a failed assertion is added to the test results.
    */
-  setup?: () => q.Promise<any>;
+  setup?(): void|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: These can actually return any kind of promise.  How about `void|q.IPromise`?  Also, update the comment to have a return type `{Promise=}`.

File: lib/plugins.ts [hunk 8]
```diff
@@ -142,14 +142,14 @@ export interface ProtractorPlugin {
...
-   *     command.
+   * @return {webdriver.promise.Promise|Q.Promise} Can return a promise, in which
+   *     case protractor will wait for the promise to resolve before continuing. If
+   *     the promise is rejected, a failed assertion is added to the test results,
+   *     and protractor will continue onto the next command. If nothing is returned
+   *     or something other than a promise is returned, protractor will continue
+   *     onto the next command.
    */
-  waitForPromise?: () => q.Promise<any>;
+  waitForPromise?(): webdriver.promise.Promise<void>|q.Promise<void>;
```
Comment (sjelin) [inline lib/plugins.ts]: `q.Promise` is actually not allowed here - only webdriver's promise implementation.  Could you fix the type/comment?

File: lib/plugins.ts [hunk 9]
```diff
@@ -162,13 +162,13 @@ export interface ProtractorPlugin {
...
-   *     assertion is added to the test results, and protractor will continue onto
-   *     the next command.
+   * @return {webdriver.promise.Promise<boolean>|Q.Promise<boolean>|boolean} If truthy,
+   *     Protractor will continue onto the next command.  If falsy, webdriver will
+   *     continuously re-run this function until it is truthy.  If a rejected promise
+   *     is returned, a failed assertion is added to the test results, and Protractor
+   *     will continue onto the next command.
    */
-  waitForCondition?: () => q.Promise<any>;
+  waitForCondition?(): webdriver.promise.Promise<boolean>|q.Promise<boolean>|boolean;
```
Comment (sjelin) [inline lib/plugins.ts]: `q.Promise` is actually not allowed here - only webdriver's promise implementation.  Could you fix the type/comment?

**Comment by cnishina** [comment]
@thorn0 ping for changes.

**Comment by sjelin** [comment]
Closing in favor of https://github.com/angular/protractor/pull/4034

**Pull request closed without merging**